utility/volume_counter/volume_counter.1
.TH VOLUME_COUNTER 1
.SH NAME
volume_counter - counts earthquakes in a GCLgrid3d object
.SH SYNOPSIS
.nf
volume_counter db out [-pf pffile -v] < infile
.SH DESCRIPTION
.LP
Large earthquake catalogs present a problem to visualization with
conventional earthquake epicenter plots let alone their 3D equivalent
because the number of points can be overwhelming and obscure details
that might be of interest.   This program takes an event->origin 
database and builds a GCLvectorfield3d volume.   The output field
is stored as a 2 vector for efficiency in the calculation.  Component 0
contains an estimate of the number of earthquakes per unit volume
at any point in a spherical shaped region surrounding that grid point.
Component 1 contains a quantity I call the fault area/volume ratio.   
(see below for the formula)
.LP
The algorithm used is fairly simple.   The program builds a GCLgrid3d
object driven by a pf control file (see below).  The base grid is 
a uniform grid mapped to earth coordinates as described in the original
GCLgrid paper by Fan et al.   At each point in this volume the program 
counts how many earthquakes in the parent db are within a specified 
radius of that point.   The count is normalized by the volume of 
the spherical region and saved at each grid point.   
In addition, it also makes an estimate of the slip area for each
event from the following relationship that can be derived between
moment magnitude and fault area:   log A = M_o - 3.296.    The program
blindly substitutes whatever magnitude is requested for M_o.   Note
this metric has the effect of essentially normalized with a b value of 1.0. 
Thus, for example, in the area calculation 10 magnitude 3s count the same
as one magnitude 4.  
.LP
Note this program eats up the entire catalog and stores it in memory 
to make the processing reasonably fast.  It is thus a fairly large 
memory algorithm if used on a large catalog over a large area with
many nodes.  Be further warned that because it has to test
every event for every node in a 3D grid, the run times can be very long 
if the grid is large and the catalog is large.   e.g. a modest grid
for the Alaska catalog took about 16 hours to run on a 2012 vintage
Mac computer.   
.LP
The first argument is required and is the database name.   The second
is the name to be assigned to the output field that will be stored 
in the db in the gclfield table.  
.SH OPTIONS
.IP -pf
Change the parameter file name.  (WARNING:  this program currently uses
a modified pf file reader that will NOT automatically append the ".pf" 
tag so this must be the full file name.)
.IP -ss
Apply next argument as a subset condition (antelope db expression) to
the event->origin join before processing. 
.IP -v
Run in verbose mode.  This is highly recommended for large catalogs
as in verbose mode a message is posted after each run through the 
fastest (z) index in the grid.
.SH PARAMETER FILE
.LP
The following are exactly as in makegclgrid(3):  delta_x1,
delta_x2, delta_x3, gridfile_directory, gridname, n1, n2, n3,
origin_latitude, origin_longitude, origin_offset_x1, origin_offset_x2,
and x_axis_azimuth.
.LP
The following relate to handling magnitudes.   The boolean 
\fIuse_netmag\fR controls handling of the netmag table.  When true
only rows of event->origin that join with netmag will be used.  
Only one magnitude will be used for area conversion.   The attribute
the program will actually try to extract is the defined by 
\fImagnitude_attribute\fR.   This parameter should use a full 
table.attribute name to avoid potentially mysterious aborts.   
The \fInull_magnitude_default\fR sets undefined magnitudes to this
value.   Normal practice should be to set this to something near the
detection floor of the network.

.SH DIAGNOSTICS
.LP
Information diagnostics that are hopefully normally clear are
sent to stdout through the C++ throw-catch mechanism.
.SH "BUGS AND CAVEATS"
.LP
This program can run a very long time on large catalogs.  If you 
do not need the full area of a catalog use a subset condition 
to reduce the size of the relavant events as this can 
dramatically reduce run times.
.SH AUTHOR
Prof. Gary L. Pavlis (pavlis@indiana.edu)


utility/volume_counter/volume_counter.cc
#include <string>
#include <fstream>
#include <list>
#include <vector>
#include "gclgrid.h"
#include "seispp.h"
#include "dbpp.h"
#include "PfStyleMetadata.h"
#include "EventCatalog.h"
using namespace std;
using namespace SEISPP;
void usage()
{
    cerr << "volume_counter db out [-ss subset_condition -pf pffile -v] < infile"<<endl
        << "   db is Antelope db  (needs event->origin and will write gcl tables"<<endl
        << "   out is name for output gclfield data (file name and in gclfield table)"<<endl
        << "   (note out is vectorfield with 0 count and 1 area/volume ration estimate)" <<endl
        << "   -ss defines subset condition applied before processing (can be important for efficiency)"<<endl
        << "   -pf for alternate pf (check pf before running of course)"<<endl
        << "   -v to turn on verbose error reporting"<<endl;;
    exit(-1);
}
class DistanceTest : public unary_function<Hypocenter,bool>
{
    public:
        DistanceTest(GCLgrid3d *gclg,int i, int j, int k,double rc)
        {
            g=gclg;
            x1p=gclg->x1[i][j][k];
            x2p=gclg->x2[i][j][k];
            x3p=gclg->x3[i][j][k];
            rcutoff=rc;
        };
        bool operator() (const Hypocenter& h) const
        {
            double ssq,delta;
            Cartesian_point cp=g->gtoc(h.lat,h.lon,r0_ellipse(h.lat)-h.z);
            delta=x1p-cp.x1;
            ssq=delta*delta;
            delta=x2p-cp.x2;
            ssq+=delta*delta;
            delta=x3p-cp.x3;
            ssq+=delta*delta;
            if(sqrt(ssq)<=rcutoff) return true;
            return false;
        };
    private:
        double x1p,x2p,x3p;
        /* This is the cutoff distance */
        double rcutoff;
        /* A pointer for efficiency.   Would be a horrible
           overhead otherwise. */
        GCLgrid3d *g;
};

double sum_areas(auto_ptr<EventCatalog> events,string magname,double nullvalue)
{
    try {
        double sumresult(0.0);
        events->rewind();
        int nevents=events->size();
        int i;
        /* EventCatalog has an internal counter - odd increment operator 
           is a syntatic oddity of usign an auto_ptr */
        for(i=0;i<nevents;++i,++(*events))
        {
            Metadata aux=events->current_aux();
            double mag=aux.get_double(magname);
            /* assume anything negative is null */
            if(mag<0.0) mag=nullvalue;
            /* glp see momentscalingtheory.pdf for derivation of this formula*/
            double area=pow(10.0,(mag-3.296));
            sumresult += area;
        }
        return sumresult;
    }catch(...){throw;};
}
bool SEISPP::SEISPP_verbose(false);
int main(int argc, char **argv)
{
    if(argc<3) usage();
    string dbname(argv[1]);
    string outfield(argv[2]);
    string pffile("volume_counter");
    bool apply_subset(false);
    string subset_string;
    int i,j,k;
    for(i=3;i<argc;++i)
    {
        string sarg(argv[i]);
        if(sarg=="-pf")
        {
            ++i;
            if(i>=argc) usage();
            pffile=string(argv[i]);
        }
        else if(sarg=="-ss")
        {
            ++i;
            if(i>=argc) usage();
            apply_subset=true;
            subset_string=sarg;
        }
        else if(sarg=="-v")
            SEISPP_verbose=true;
        else
            usage();
    }
    pffile=pffile+".pf";
    ios::sync_with_stdio();
    try{
        DatascopeHandle dbh(dbname,false);
        if(apply_subset) 
        {
            if(SEISPP_verbose)
                cout << "Original db size before subset application"
                    << dbh.number_tuples()<<endl
                    << "Applying this subset condition:  "
                    << apply_subset<<endl;
            dbh.subset(subset_string);
            if(SEISPP_verbose)
                cout << "Number of tuples after subset = "
                    << dbh.number_tuples()<<endl;
        }
        PfStyleMetadata control=SEISPP::pfread(pffile);
        if(SEISPP_verbose)
        {
            cout << "Parameter file values read"<<endl;
            control.pfwrite(cout);
        }
        double search_radius=control.get_double("search_radius");   // this is in km
        string dir=control.get_string("output_directory");
        /* This is essential the same as makegclgrid but using the new PfStyleMetadata object */
        double lat0,lon0,r0,azimuth_x,dx1,dx2,dx3;
        int n1,n2,n3,i0,j0;
        lat0=control.get_double("origin_latitude");
        lon0=control.get_double("origin_longitude");
        lat0=rad(lat0);
        lon0=rad(lon0);
        r0=r0_ellipse(lat0);
        azimuth_x=control.get_double("x_axis_azimuth");
        double azimuth_y=azimuth_x-90.0;
        azimuth_y=rad(azimuth_y);
        dx1=control.get_double("delta_x1");
        dx2=control.get_double("delta_x2");
        dx3=control.get_double("delta_x3");
        n1=control.get_int("n1");
        n2=control.get_int("n2");
        n3=control.get_int("n3");
        i0=control.get_int("origin_offset_x1");
        j0=control.get_int("origin_offset_x2");
        string gridname=control.get_string("gridname");
        bool use_netmag=control.get_bool("use_netmag");  
        string magnitude_attribute=control.get_string("magnitude_attribute");
        double null_magnitude_default=control.get_double("null_magnitude_default");
        if(use_netmag)
        {
            if(SEISPP_verbose)
                cout << "Attempting to join netmag table"<<endl
                    << "Current view size before join="<<dbh.number_tuples()
                    <<endl;;
            dbh.natural_join(string("netmag"));
            if(SEISPP_verbose)
                cout << "Number of rows after joining netmag="
                    << dbh.number_tuples()<<endl;
        }
        /* We always load ms, mb, and ml as they are part of origin*/
        MetadataList maglist;
        Metadata_typedef md;
        md.tag=string("origin.ms");
        md.mdt=MDreal;
        maglist.push_back(md);
        md.tag=string("origin.mb");
        maglist.push_back(md);
        md.tag=string("origin.ml");
        maglist.push_back(md);
        if(use_netmag)
        {
            md.tag=string("netmag.magnitude");
            maglist.push_back(md);
        }
        AttributeMap am("css3.0");
        EventCatalog eqs(dbh,maglist,am);
        /* Use new here so we can save some memory once
           we create the field object */
        GCLgrid3d *g=new GCLgrid3d(n1,n2,n3,gridname,lat0,lon0,r0,
                azimuth_y,dx1,dx2,dx3,i0,j0);
        GCLvectorfield3d f(*g,2);
        /* this is not essential, but best to be explicit */
        f.zero();
        delete g;

        /* Now loop through all the points and count */
        const double volume(4.0*M_PI*search_radius*search_radius*search_radius/3.0 );
        list<Cartesian_point>::iterator points;
        for(i=0;i<f.n1;++i)
            for(j=0;j<f.n2;++j)
            {
                if(SEISPP_verbose)
                        cout<<"Starting on index i,j="<<i<<" "<<j<<endl
                            << "Time is "<<strtime(now())<<endl;
                for(k=0;k<f.n3;++k)
                {
                    DistanceTest closeones(&f,i,j,k,search_radius);
                    auto_ptr<EventCatalog> locals=eqs.subset<DistanceTest>(closeones);
                    int count=locals->size();
                    double asum=sum_areas(locals,magnitude_attribute,null_magnitude_default);
                    /* Note we convert to count per unit volume */
                    f.val[i][j][k][0]=static_cast<double>(count)/volume;
                    f.val[i][j][k][1]=asum/volume;
                }
            }
        /* This always writes the field output data to a file set by the name outfield */
        f.save(dbh,dir,dir,outfield,outfield);
    }catch(std::exception& err)
    {
        cerr << "Something threw the following error:"<<endl
            << err.what()<<endl;
    }
    catch(SeisppError& serr)
    {
        serr.log_error();
    }
}






